### ISSUE TITLE
Rescuing timeout errors in middleware

### ISSUE THREAD

**Issue by chanks** [heroku/rack-timeout#76]
Hi - I've been running into a problem recently with Rack::Timeout's errors killing Puma workers (since they now descend from Exception rather than StandardError and are not caught by most rescue blocks).

I've worked around this with a quick rack middleware to catch Rack::Timeout's errors and wrap them in StandardErrors:

```
class TimeoutRecovery
  def initialize(app, options = {})
    @app, @options = app, options
  end

  # Rack::Timeout, as of 0.2.4, raises errors that descend from the Exception
  # class when a request takes too long - however, this is a very strong
  # exception level and will kill Puma's worker if we let it continue down the
  # stack. So instead, catch it and wrap it in a recoverable StandardError.

  class Error < StandardError; end

  def call(env)
    @app.call(env)
  rescue Rack::Timeout::Error => e
    raise Error, "Original Error: #{e.class.inspect}, #{e.message.inspect}, #{e.backtrace.inspect}"
  end
end
```

It seems like this is something that Rack::Timeout should be doing, though. Unless I've missed something and this is actually a terrible idea. Sure, we should be avoiding long request times where possible, but when they come up they're leaving Puma's workers crashed and unable to serve requests at all until a restart.

**Comment by kch** [comment]
@keiko713 wonder if it relates to your thing

**Comment by kch** [comment]
@chanks what framework are you using? or plain rack? rails and sinatra catch all sorts of exceptions…

This change was made as a brute-force solution to bad codez rescuing from StandardError (rescue without a parameter) in people's apps.

**Comment by chanks** [comment]
This app is using Rails, but I imagine it'd be much the same with any stack. I think having it descend from Exception is reasonable to avoid the blanket-rescue problem, but that Rack::Timeout should probably then be rescuing Rack::Timeout::Error itself and then reraising something that Puma (or, I would imagine, other servers) can recover from.

**Comment by kch** [comment]
I thought rails would rescue from exception at some point… maybe they got sane and don't anymore? Haven't touched the thing much since ~2.3.

Yeah, catch and re-raise sounds ok in that light, but I may need wiser people to chime in. @watsonian @chadbailey59 opinions?

**Comment by chanks** [comment]
Yes, I generally prefer Sinatra (or Roda, these days), but what can you do.

BTW, the readme should be updated too, it explicitly says that Rack::Timeout's errors descend from RuntimeError.

**Comment by keiko713** [comment]
Interesting. @chanks , can you please share me what kind of errors you see when Rack::Timeout killed Puma workers? It may be different from my thing if it started from v2.4 though.

**Comment by chanks** [comment]
We were running Puma on Heroku (in single process mode), and we would occasionally experience a large surge in traffic that would overload our dynos. Once it subsided, no requests would be serviced, everything would just fail after 30 seconds with an H12 error until the app was restarted.

I debugged it by deploying the app to staging with a SIGTERM handler that would cycle through all threads via ObjectSpace, join them, and rescue/log the exceptions that had killed them. Then I'd just overload the staging app with traffic until it became unresponsive and then do a heroku ps:restart. This was how I discovered that all of Puma's worker threads had died due to Rack::Timeout::Errors. This is all on rack-timeout 0.2.0 and puma 2.11.1.

**Comment by keiko713** [comment]
That sounds very much like my thing. Thanks for explaining.

**Comment by kch** [comment]
I wonder if using `throw/catch` along with / instead of `raise/rescue` might help. Haven't given it much thought, but leaving it here to look into later.

**Comment by chadbailey59** [comment]
I can definitely see arguments from both sides—since rack-timeout kind of exists outside the context of an entire Rails app, you wouldn't want Rails to be responsible for handling its exceptions (or unintentionally ignoring them by bulldozing `StandardError`s).

However, it seems the consequences for a generic `Exception` might be too strong. If it's causing problems with Puma, it could be causing similar problems with Unicorn. We've been tracking an issue where unicorn workers will die without closing database connections; I wonder if something similar is happening?

**Comment by kch** [comment]
> We've been tracking an issue where unicorn workers will die without closing database connections; I wonder if something similar is happening?

Yeah but that's a known possibility and the solution to that is similar. Add on observer or rescue the error, close the connections, then probably re-raise.

I don't know if there's a one-size-fits-all solution here, and more and more I'm leaning against raising at all. Just provide a hook for people to do whatever they want with it? Then maybe provide some standard behaviours: log, raise, die… which I suppose one could do with observers already… 

![250px-054psyduck](https://cloud.githubusercontent.com/assets/8216/7261876/046ae5d0-e82a-11e4-91a4-bff60fd73ab8.png)

**Comment by chadbailey59** [comment]
What if we left the default behavior as-is, but had a simple DSL to
override it in an initializer? If you don’t do anything, it will raise an
Exception when it hits the timeout, but you could do something like this:


```

Rack::Timeout.timeout = 5

Rack::Timeout.warning = 2


on :warning do

  # start spitting out a backtrace here

end


on :timeout do

  # raise an Exception or StandardError here

end

```

**Comment by chadbailey59** [comment]
What if we left the default behavior as-is, but had a simple DSL to override it in an initializer? If you don’t do anything, it will raise an Exception when it hits the timeout, but you could do something like this:

```ruby
Rack::Timeout.timeout = 5
Rack::Timeout.warning = 2

on :warning do
  # start spitting out a backtrace here
end

on :timeout do
  # raise an Exception or StandardError here
end
```

**Comment by chanks** [comment]
Sorry, you all know Rack::Timeout's internals better than I do, but what's the issue with the solution I proposed, of catching the Rack::Timeout error in the middleware and then reraising an error that descends from StandardError and is therefore more recoverable? Basically what I did in the snippet in the initial issue. I think it handles the problem with blanket rescues without running the risk of killing the thread/process.

**Comment by chanks** [comment]
(I ask partly because it seems like the simplest solution, and partly because if that snippet I'm running in production is dangerous, I'd like to know about it :smile:)

**Comment by kch** [comment]
I think your snippet is pretty safe and I don't see any problems with your solution, other than it's piling up even more code on a thing that should be simple. It may well be that that's how we address it. Just taking the time to see if there's another design change that might be a better approach.

**Comment by kch** [comment]
(That said, running rack-timeout at all is dangerous, as explained in the readme :smile: )

**Comment by chanks** [comment]
Ha, alright, good to remember!

**Comment by travisp** [comment]
It's dangerous, but running rack-timeout is part of what is typically recommended for setup of puma at heroku:

https://devcenter.heroku.com/articles/deploying-rails-applications-with-the-puma-web-server#timeout

Personally, I think this middleware code should probably be part of the heroku doc recommendations until there is an alternate way to handle, or a warning that, left alone, rack-timeout probably shouldn't be used on a production app using puma. It was the only thing that explained/fixed why we were finding dead puma workers on occasion. While we do what we can to avoid ever hitting a rack-timeout, it's difficult to prevent from ever happening if the servers get overwhelmed.

**Comment by zsalzbank** [comment]
I was having issues getting the middleware into the right position using rails, so I made a gem that does it for me: https://github.com/keyme/rack-timeout-puma.

**Comment by kch** [comment]
I'll make this part of rack-timeout soon.

**Comment by ankane** [comment]
`Rack::Timeout::RequestExpiryError` kills Unicorn as well (creates a bad spiral) since it's not rescued, so we rearrange the middleware in `config/application.rb`.

```ruby
app.config.middleware.delete "Rack::Timeout"
app.config.middleware.insert_before "ActionDispatch::RemoteIp", "Rack::Timeout"
```

This has the added benefit of sending the error to reporting services and showing a custom error page.

One potential downside is timeouts won't happen in earlier middleware, but the Unicorn timeout handles this situation.

**Comment by kch** [comment]
The solution is to rescue like the OP said, and yeah, when I get time I'll add this in.

**Comment by kch** [comment]
Although yeah getting the benefit of custom error pages etc is good. I've thought about responding with a 503… the problem is each framework handles things differently, so ideally one rescues and re-raises whatever they want.

**Comment by ankane** [comment]
Yeah, Rails will take care of the rescue if inserted after `ActionDispatch::ShowExceptions` (unsure how other Rack-based frameworks handle it).

**Comment by vincentwoo** [comment]
The latest version of puma has this line in the changelog:

`  * Don't allow any exceptions to terminate a thread`

Does that change anything about this situation?

**Comment by travisp** [comment]
For reference, here is the change:

https://github.com/puma/puma/commit/0d8f026172bf051fcd60f698b9ff4a9f893c9ef4

Puma has also added thread reaping.

**Comment by kch** [comment]
I'd expect it to make the extra middleware no longer necessary, but haven't really tested to confirm.

Is all that in a public release yet? I saw the original commit being made but haven't looked since for releases.

**Comment by vincentwoo** [comment]
It is! https://github.com/puma/puma/blob/master/History.txt

**Comment by kch** [comment]
Cool, so this issue can probably be closed. I'll keep it around until I do the next round of work on r-t.

**Comment by vincentwoo** [comment]
FWIW, I find myself still having to use puma-rack-timeout, otherwise (even on the newest puma) I see problems.

**Comment by kch** [comment]
Issues about each particular problem would be welcome :)

**Comment by vincentwoo** [comment]
I don't have too formal a report, but I think I was seeing a lot of H18 errors, and having rack-timeout's errors work in this way made the problems very hard to diagnose. Will let you know if I have something more specific.

**Comment by jeremyhaile** [comment]
We were just running into this issue. Our application would timeout when under initial load, which would cause TimeoutErrors to occur. These would then cause puma threads to die, which would never be re-spawned. Upgrading to the latest puma (2.12.2) and switching to rack-timeout-puma has fixed the issue in my tests.

**Comment by kch** [comment]
Should be resolved by #86
